nni/compression/pytorch/speedup/jit_translate.py

```diff
@@ -462,178 +164,295 @@ def forward(self, obj):
-    TODO: support more types in the type_as, need to figure out
-    how to get the scalar type from torch._C.TensorType.
+    if ivalue is None or type(ivalue) is torch.dtype:
+        return ivalue
+    elif type(ivalue) is int:
+        global enum2dtype_dict
+        if ivalue not in enum2dtype_dict:
+            raise TypeError("Unimplemented scalar type")
+        return enum2dtype_dict[ivalue]
+    else:
+        raise TypeError("Unimplemented scalar type")
+
+enum2memory_format_dict = {
+    0: torch.contiguous_format,
+    1: torch.preserve_format, # char
+    2: torch.channels_last,
+    3: torch.channels_last_3d,
+}
+def memory_format_trans(ivalue: int | torch.memory_format):
     """
-    class TypeasModule(torch.nn.Module):
-        def __init__(self, dtype=torch.float):
-            self.example = torch.zeros(1, dtype=dtype)
+    Special process for memory_format.
+    Torch will transform memory_format to an enum in cpp, so the value of memory_format we get in jit is an int.
+    This function is used to recover the int to torch.memory_format in python.
 
-        def forward(self, x):
-            return x.type_as(self.example)
-    return TypeasModule()
+    Parameters
+    ----------
+    ivalue:
+        The value of memory_format or method to be recovered.
 
+    """
+    if ivalue is None or type(ivalue) is torch.memory_format:
+        return ivalue
+    elif type(ivalue) is int:
+        global enum2memory_format_dict
+        if ivalue not in enum2memory_format_dict:
+            raise TypeError("Unimplemented memory_format type")
+        return enum2memory_format_dict[ivalue]
+    else:
+        raise TypeError("Unimplemented memory_format type")
 
-def to_python(node, speedup):
-    # for the time being, only device parameters are supported
-    class ToModule(torch.nn.Module):
-        def __init__(self, device):
-            super(ToModule, self).__init__()
+special_treat_dict = {
+    'dtype': dtype_trans,
+    'memory_format': memory_format_trans,
+}
+def generate_aten_to_python(func: Callable, node: NodePyGroup, speedup: ModelSpeedup):
+    """
+    parse a
+    Return a callable object to inference the mask according to the
+    node.op_type.
 
-        def forward(self, x):
-            return x.to(device)
+    Parameters
+    ---------
+    func: Callable
+        The torch function one-to-one correspondence with the node.
+    node: NodePyGroup
+        The target node to inference the mask
+    speedup: ModelSpeedup
+        The speedup object of the target model.
 
+    Returns
+    ------
+    func: callable object(nn.Module/function)
+        Return the translated function that used to inference the mask
+        , if current op_type is not supported, then we return None.
+    """
     c_node = node.key_node
-    inputs = list(c_node.inputs())
-    device = inputs[3].toIValue()
-    return ToModule(device)
+    schema = c_node.schema()
+
+    ## parse the schema, to positional_num and keyword_list
```

Comment:
Please put the parsing logic into a separate function. It's complex.

Reply:
ok